### PR TITLE
Movie Origin / Domain Lock Pass-Through

### PR DESCRIPTION
**Pull request by kmeisthax**
Currently, we trip every domain lock in every movie we run, as per #325 (which also crops up in #751). This is because we are not storing the origin URL of the SWFs we are loading. This PR moves movie loading into the player core and stores an origin URL on every `SwfMovie` which the movie's ActionScript can read to verify the domain it's hosted on.

Care was placed to avoid breaking the demo site, which relies on being able to run movie data directly. However, no attempt will be made to allow the demo site to falsify domain data as this would make it pretty trivial to break domain locks.

The work in this PR may also be of use for #379 as the `StorageBackend` could use the movie URL (or a hash of it) as a namespace for the keys.

- [x] Add an origin URL parameter to `SwfMovie`
- [x] Allow Ruffle to load the movie itself
- [x] Add backend method for resolving relative paths
- [x] Ensure all Ruffle-fetched SWFs have resolved paths
- [x] Ensure all caller-provided movies have resolved paths
- [x] Implement `_url` for display objects

### CODE REVIEW

**Comment by Herschel** [comment]
We should get rid of `play_swf_data` on the web side and use `stream_swf_url` on the demo as well.

**Comment by Dinnerbone** [comment]
Does that work for files provided by the user?

**Comment by Herschel** [comment]
Oh, that's true -- we'll still need `play_swf_data` for the local files for now. `loadRemoteFile` in the demo should use `stream_swf_url` at least. When we actually implement streaming, maybe the player could take a more generic future to handle different stream types (can you make a ReadableStream out of a FileReader?)

**Comment by kmeisthax** [comment]
Releasing the draft lock on this now that I've managed to successfully pass domain locks that use `_url`. We generate valid URLs on the web, while desktop gets file URLs.

I've deliberately designed the JavaScript API to mildly inconvenience people attempting to use Ruffle to breach domain locks. SWF data piped directly to the player (e.g. from a file upload) is given no origin data, so no domain locks can be bypassed by just calling `play_swf_data`. I (naively) figure the people who could figure out how to bypass this very meager protection would probably be smart enough to not waste their time stealing old Flash movies. Of course, if the player has fetched the movie itself, then it gets an origin URL and all domain locks will pass.

File: core/src/backend/navigator.rs [hunk 1]
```diff
@@ -11,6 +12,39 @@ use std::sync::mpsc::{channel, Receiver, Sender};
...
+    let parsed = Url::from_file_path(relative);
+    if let Err(()) = parsed {
+        let base =
+            Url::from_directory_path(base).map_err(|_| ParseError::RelativeUrlWithoutBase)?;
+
+        return base.join(relative);
+    }
+
+    Ok(parsed.unwrap())
+}
```
Comment (Herschel) [inline core/src/backend/navigator.rs:33]: Can we add the `#[cfg(not(...))]` variant here that just returns an error or returns the URL as-is, and then we can remove the conditionals that are elsewhere?

**Pull request merged**